src/main/java/io/stargate/sgv2/jsonapi/service/updater/DocumentUpdater.java

```diff
@@ -4,29 +4,111 @@
-public record DocumentUpdater(List<UpdateOperation> updateOperations) {
+public record DocumentUpdater(
+    List<UpdateOperation> updateOperations,
+    ObjectNode replaceDocument,
+    JsonNode replaceDocumentId,
+    UpdateType updateType) {
+  /**
+   * Construct to create updater using update clause
+   *
+   * @param updateDef
+   * @return
+   */
   public static DocumentUpdater construct(UpdateClause updateDef) {
-    return new DocumentUpdater(updateDef.buildOperations());
+    return new DocumentUpdater(updateDef.buildOperations(), null, null, UpdateType.UPDATE);
+  }
+
+  /**
+   * Construct to create updater using replace document
+   *
+   * @param replaceDocument
+   * @return
+   */
+  public static DocumentUpdater construct(ObjectNode replaceDocument) {
+    JsonNode replaceDocumentId = replaceDocument.remove(DocumentConstants.Fields.DOC_ID);
+    return new DocumentUpdater(null, replaceDocument, replaceDocumentId, UpdateType.REPLACE);
   }
 
   /**
    * @param readDocument Document to update
    * @param docInserted True if document was just created (inserted); false if updating existing
    *     document
    */
-  public DocumentUpdaterResponse applyUpdates(JsonNode readDocument, boolean docInserted) {
+  public DocumentUpdaterResponse apply(JsonNode readDocument, boolean docInserted) {
     ObjectNode docToUpdate = (ObjectNode) readDocument;
+    if (UpdateType.UPDATE == updateType) {
+      boolean modified = update(docToUpdate, docInserted);
+      return new DocumentUpdaterResponse(readDocument, modified);
+    } else {
+      boolean modified = replace(docToUpdate, docInserted);
+      return new DocumentUpdaterResponse(readDocument, modified);
+    }
+  }
+
+  /**
+   * Will be used for update commands
+   *
+   * @param docToUpdate
+   * @param docInserted
+   * @return
+   */
+  private boolean update(ObjectNode docToUpdate, boolean docInserted) {
     boolean modified = false;
     for (UpdateOperation updateOperation : updateOperations) {
       if (updateOperation.shouldApplyIf(docInserted)) {
         modified |= updateOperation.updateDocument(docToUpdate);
       }
     }
-    return new DocumentUpdaterResponse(readDocument, modified);
+    return modified;
+  }
+
+  /**
+   * Will be used for findOneAndReplace
+   *
+   * @param docToUpdate
+   * @param docInserted
+   * @return
+   */
+  private boolean replace(ObjectNode docToUpdate, boolean docInserted) {
+    // replace will be false if upsert, should never happen
+    if (docInserted) {
+      return false;
+    }
+    // Do deep clone so we can remove _id field and check
+    ObjectNode compareDoc = docToUpdate.deepCopy();
+    JsonNode idNode = compareDoc.remove(DocumentConstants.Fields.DOC_ID);
+    // The replace document cannot specify an _id value that differs from the replaced document.
+    if (replaceDocumentId != null) {
+      if (JsonNodeComparator.ascending().compare(replaceDocumentId, idNode) != 0) {
```

Comment:
Simple equality check works for kinds of `JsonNode`s that are legal as doc-ids, so no need for comparator.

But maybe better is to use existing method from `JsonUtil`; `equalsOrdered()`? That would work even if allowed more types and is designed for equality checks.

Reply:
OK. Will use JsonUtil